Initial acute-eval open source commit

hadasah opened the pull request:
**Patch description**
Open sourcing ACUTE-EVAL 

worlds.py needs cleanup (as does run.py still) but putting out PR for comments

stephenroller commented:
That was a lot less code than it felt like at the time :D

hadasah commented:
I mean, I've deleted *many* lines over the many iterations lol.

emilydinan commented on parlai/mturk/tasks/acute_eval/example_script.py:
```diff
@@ -0,0 +1,54 @@
+
+def set_args():
+    args = add_args()
+    args['dialogs_path'] = '/home/ParlAI/parlai/mturk/tasks/acute_eval/example'
```
can we set this from the person's parlai home variable? it might not be in /home/ParlAI

hadasah replied:
this was only meant as an example - I expect people to copy this script and modify all of the args

mwillwork replied:
Could this be another command line param? (And put in the README the format it expects the file to be in?) It would be nice also to specify what it does if you give it a directory versus what it does when you get it a file and have the default behavior do the right. In addition, one nice thing would be if someone just has the ability to implement a single function that parses their file. (So pull this out into its own function that can be overridden:  https://github.com/facebookresearch/ParlAI/blob/8a3018dc8e2f4f5f6992c2a7058c27fb267b1524/parlai/mturk/tasks/acute_eval/run.py#L163 ) The frontend expects certain fields to be there such as "dialog",  so the format should be specified to.

mwillwork replied:
Could this be another command line param? (And put in the README the format it expects the file to be in?)

stephenroller replied:
It actually is a CLI arg, Margaret chose to set it via a script.

emilydinan commented on parlai/mturk/tasks/acute_eval/example_script.py:
```diff
@@ -0,0 +1,54 @@
+"""
+Example script for running ACUTE-EVAL.
+The only arguments that need to be modified for this to be run are:
+'dialogs_path':  Path to folder containing logs of all model conversations.
```
might be a bit more natural if we made these arguments to `set_args` instead of having to hard code the values in this script

hadasah replied:
I set it up to force users to write a script because we have a lot of really long strings and lists in the arguments, and it would be really ugly to try to parse it from commandline, but open to suggestions

emilydinan replied:
I meant you could set as function args in `set_args()` not as CLI args. that way you can just import this function and switch the args instead of having to re-write the body of the function every time.

hadasah replied:
I see. I don't particularly have a problem with that, but it would be no different from what currently exists, just in a slightly different place. In that sense, I feel like it's actually cleaner to have it this way if you do want to set a lot of these args to custom (non-default) values.

emilydinan commented on parlai/mturk/tasks/acute_eval/run.py:
```diff
@@ -0,0 +1,544 @@
+    """ Initialize task queue to contain the specified number of instances of
+    each pairing
+    """
+    # hacky fix for the parlai parser hacky fix
```
hmm... why aren't the hyphens already replaced?

hadasah replied:
it wasn't at the time I wrote this - I think the hyphen replacing was happening before my args were being added, but then the underscores in my args got converted to hyphens - maybe it's been fixed since.

emilydinan commented on parlai/mturk/tasks/acute_eval/run.py:
```diff
@@ -0,0 +1,544 @@
+    for data_fn in os.listdir(data_folder):
+        full_data_fn = os.path.join(data_folder, data_fn)
+        with open(full_data_fn, 'r') as dialog_data_file:
+            model_name = data_fn.split('.')[0]
```
so the data is expected to be in the data folder with specific file formatting? can we document this somewhere?

hadasah replied:
yes, I need to write a good README

emilydinan replied:
what if instead we made this *not* dependent on the datafile name? shouldn't the model name be specified in the json anyway? can't we just provide some root dir to search through for json's with that model name?

hadasah replied:
I would be ok with instead specifying a dict with model_name: file_name - is that sort of what you had in mind?

emilydinan replied:
I was thinking more along the lines of: each JSON conversation has the key `model_name`, right? Can we just specify the root directory, search all conversations, and select by `model_name` key?

mwillwork replied:
It would be nice to have default functionality that does the right thing if you give it a directory and also if you give it a file. In addition, one nice thing would be if someone just has the ability to implement a single function that parses their file. (So pull this out into its own function that can be overridden.) I have to parse a CSV file someone else gave me for example. The frontend expects certain fields to be there such as "dialog",  so the format should be specified too as per emily's comment above.

emilydinan commented on parlai/mturk/tasks/acute_eval/run.py:
```diff
@@ -0,0 +1,544 @@
+                    subtask_data['conversations'][1]
+                )
+            except ValueError():
+                print("WARNING: couldn't remove task from worker's history")
```
when would this happen?

hadasah replied:
in theory, never, and I never saw this triggered.  But apparently it's technically possible for weird race conditions to happen with MTurk, so I left some of these in just in case.

emilydinan commented on parlai/mturk/tasks/acute_eval/run.py:
```diff
@@ -0,0 +1,544 @@
+
+
+def get_onboarding_tasks(worker_id, tasks_per_hit):
+    """ Get the next onboarding task for this worker id. If the worker has never
```
can you add the argument descriptors to your docstrings?

emilydinan commented on parlai/mturk/tasks/acute_eval/run.py:
```diff
@@ -0,0 +1,544 @@
+        '--pair_data',
+        type=list,
+        default=None,
+        help='list of (conv1, conv2, hit, desc) (for pre-chosen pairs, e.g. for replicating previous experiments)',
```
also, what is hit and desc ion this argument?

hadasah commented on parlai/mturk/tasks/acute_eval/example_script.py:
```diff
@@ -0,0 +1,54 @@
+    args['max_hits_per_worker'] = 1
+
+    # Additional args that can be set - here we show the default values.
+    # For a full list, refer to run.py & the parlai/params.py
```
No, they don't need to be, but since I used all of the default values, there's no difference in the function. I can uncomment if that looks better.

klshuster replied:
yes please 😄

hadasah commented on parlai/mturk/tasks/acute_eval/run.py:
```diff
@@ -0,0 +1,544 @@
+            task_queue.put(desired_tasks[internal_id])
+    # limit number of hits worker can do by default
+    if opt['max_hits_per_worker'] == 0:
+        opt['max_hits_per_worker'] = (len(desired_tasks) + len(onboarding_tasks)) / opt[
```
yes, good catch.

stephenroller replied:
use // to avoid the float

JackUrb commented:
Took a quick pass over the react, nothing seems out of place to me

klshuster commented on parlai/mturk/tasks/acute_eval/example_script.py:
```diff
@@ -0,0 +1,54 @@
+from parlai.mturk.tasks.pairwise_dialogue_eval.run import main as run_main, add_args
```
also a quick note that this is the reason the PR is failing all tests currently lol

mwillwork commented:
A few suggestions:
1.  One nice thing would be if someone just has the ability to implement a single function that parses their file and then everything else just works. (So pull this out into its own function that can be overridden:  https://github.com/facebookresearch/ParlAI/blob/8a3018dc8e2f4f5f6992c2a7058c27fb267b1524/parlai/mturk/tasks/acute_eval/run.py#L163 ) The frontend expects certain fields to be there such as "dialog",  so the format should be specified to in documentation.
2. As discussed in person, for checking the onboarding tasks it would be nice if it didn't always assume the 2nd answer is correct. (i.e. have the ability to give an expected answer in the JSON.) 
3. I don't want to use any of my existing conversations as my onboarding tasks, so I'm going to put some custom JSON and create a function to build the onboarding tasks from that. Would be good to support this.
4. Can you put in the README that you are using "task" intentionally - and what a comparison is versus a pair versus a task versus a conversation versus a dialog. This would make the code easier to figure out.  
5. I think this functionality probably belongs in a class (call it ACUTEEvalBuilder?): you have a bunch of functions and some global variables, so basically this functionality all is related and shares state...hence class.
6. Would be good to have tests for this if it's going to be our standard eval method. I did a related MTurk task where I heavily modified this code (so it could have been my new code), but I somehow ended up doing more annotations per task than I wanted and it's not easy to figure out why. A test testing various flags, workers getting blocked, how the tasks are selected, incomplete tasks being returned to the task queue, replicating a previous run, etc would be great.

emilydinan commented on parlai/mturk/tasks/acute_eval/run.py:
```diff
@@ -0,0 +1,544 @@
+            workers[0].id = display_agent_name
+
+        def run_conversation(mturk_manager, opt, workers):
+            task_data = get_new_task_data(workers[0], opt['comparisons_per_hit'])
```
`comparisons_per_hit` is not an argument above, so this causes an error. did you mean `subtasks` per hit?

emilydinan commented on parlai/mturk/tasks/acute_eval/run.py:
```diff
@@ -0,0 +1,544 @@
+    change their onboarding status depending on the task"""
+    for subtask_data in task_data:
+        if subtask_data['task_specs'].get('is_onboarding', False):
+            worker_onboarding_tasks_todo[worker_id].append(
```
this is not defined. i think you meant `workers_to_onboarding_tasks_todo`. causes an error.

emilydinan commented on parlai/mturk/tasks/acute_eval/run.py:
```diff
@@ -0,0 +1,544 @@
+    worker_id = worker.worker_id
+    task_data = get_onboarding_tasks(worker_id, tasks_per_hit)
+    if len(task_data) == tasks_per_hit:
+        return onboarding_tasks
```
this will also break here -- did you mean to return `task_data`

emilydinan commented on parlai/mturk/tasks/acute_eval/example_script.py:
```diff
@@ -0,0 +1,54 @@
+    args['model_comparisons'] = [('modela', 'modelb')]
+    args['onboarding_tasks'] = [('ZYX', 'AGHIJK', 'example_qual')]
+    args['task_description'] = {
+        'num_subtasks': 2,
```
i think this should be the existing arg, `subtasks_per_hit`, instead of `num_subtasks`.

emilydinan replied:
EDIT: this is why i found that issue with the Optional feedback not submitting.

stephenroller commented on parlai/mturk/tasks/acute_eval/frontend/components/custom.jsx:
```diff
@@ -0,0 +1,514 @@
+                    <Col sm={6}>
+                        <Radio
+                          name="speakerChoice"
+                          value="1"
```
it'd be better if these are the model names, so what we dump to disk later is more easily analyzed.

stephenroller commented on parlai/mturk/tasks/acute_eval/frontend/components/custom.jsx:
```diff
@@ -0,0 +1,514 @@
+                          name="speakerChoice"
+                          value="1"
+                          style={{'width': '100%'}}
+                          checked={this.state.speakerChoice == "1"}
```
this would have to be changed too

stephenroller commented on parlai/mturk/tasks/acute_eval/frontend/components/custom.jsx:
```diff
@@ -0,0 +1,514 @@
+                      <Col sm={6}>
+                        <Radio
+                          name="speakerChoice"
+                          value="2"
```
ditto

stephenroller commented on parlai/mturk/tasks/acute_eval/frontend/components/custom.jsx:
```diff
@@ -0,0 +1,514 @@
+                          name="speakerChoice"
+                          value="2"
+                          style={{'width': '100%'}}
+                          checked={this.state.speakerChoice == "2"}
```
ditto

hadasah commented on parlai/mturk/tasks/acute_eval/run.py:
```diff
@@ -0,0 +1,544 @@
+    while (not task_queue.empty()) and tries < task_queue.qsize():
+        try:
+            next_task = task_queue.get()
+        except Queue.Empty:
```
queues technically aren't thread safe. In theory, you could have some weird cases where the queue is waiting for some items to be put back, or where qsize is still.

hadasah commented on parlai/mturk/tasks/acute_eval/run.py:
```diff
@@ -0,0 +1,544 @@
+    question,
+    is_flipped,
+    hitid='',
+    matchup='regular',
```
I don't really need the matchup field to exist at all, 'regular' was just a placeholder value. I was mostly using 'matchup' to mark qual comparisons

hadasah commented on parlai/mturk/tasks/acute_eval/run.py:
```diff
@@ -0,0 +1,544 @@
+            )
+            internal_id += 1
+    else:
+        raise NotImplementedError("--onboarding_tasks cannot be empty or None")
```
changed this to warn_once

hadasah commented on parlai/mturk/tasks/acute_eval/run.py:
```diff
@@ -0,0 +1,544 @@
+        '--onboarding_tasks',
+        type=list,
+        default=None,
+        help='onboarding tasks to screen workers with, list of (conv1id, conv2id, matchup) tuples',
```
matchup is actually never used for anything in run.py. I mostly use it to track qual annotations and what models are being compared. Not sure whether I want to keep matchup here but if I do, I'll make a note in the readme

hadasah commented on parlai/mturk/tasks/acute_eval/example_script.py:
```diff
@@ -0,0 +1,54 @@
+'model_comparisons': List of tuples indicating pairs of models to be compared.
+    The model names here should match the names of files in the folder
+'onboarding_tasks': List of tuples in format (id1, id2, name of matchup) where
+    id1, id2 are conversation ids where id2 is the correct conversation's id
```
added something to run.py, will add something to readme as well

stephenroller commented on parlai/mturk/tasks/acute_eval/README.md:
```diff
@@ -0,0 +1,108 @@
+following BibTex entry:
+
+  @misc{li2019acuteeval,
+    title={ACUTE-EVAL: Improved Dialogue Evaluation with Optimized Questions and Multi-turn Comparisons},
```
Why not the workshop?

hadasah replied:
It's not archival? Does it still have proceedings that can be cited?

hadasah commented:
Addressing some comments:

I could potentially move the next button, but we iterated a few times to settle on the current UI, and I didn't see an unusual number of returns.

Pairing logic has been removed from run.py. I'll add a separate file that enables random pairing creation. The new run.py requires a jsonl files that explicitly defines one pairing per line.

The answer recorded is now the name of the model selected.

emilydinan commented on parlai/mturk/tasks/acute_eval/run.py:
```diff
@@ -0,0 +1,413 @@
+        help='Number of annotations per conversation comparison pair',
+    )
+    argparser.add_argument(
+        '--pairings-filename', type=str, default='pairings.jsonl', help=''
```
can you add a help string here? and elsewhere where it's missing

emilydinan commented on parlai/mturk/tasks/acute_eval/run.py:
```diff
@@ -0,0 +1,413 @@
+    Build task dict according to expected format.
+    """
+    conv_orders = [[0, 1], [1, 0]]
+    conv_order = conv_orders[np.random.choice([0, 1])]
```
nit: can get rid of these two lines and just do `conv_order = random.choice([[0, 1], [1, 0]])`

emilydinan replied:
alternatively, it seems you only really need a bool to track this conversation order (i.e. you only care about the first index)

emilydinan commented on parlai/mturk/tasks/acute_eval/run.py:
```diff
@@ -0,0 +1,413 @@
+        if worker_id in onboarding_failed_workers:
+            return_task_data(worker_id, all_task_data)
+        return
+    if (num_correct / num_onboarding_tasks) >= threshold:
```
can you add some  more comments in the code explaining the flow through this funciton?

emilydinan commented on parlai/mturk/tasks/acute_eval/run.py:
```diff
@@ -0,0 +1,413 @@
+
+        sys.path.append(opt['task_folder'])
+        from task_config import task_config
+    except Exception:
```
when does this happen? should we add a warning?

hadasah replied:
We have a default task_config with the title, description, tags of the task to be shown on MTurk, but if the user writes one, I try to load it. I decided to go with having them write it in a .py file instead of .json (which would be easier to deal with) because I kept Jack's task_config.py template. There are a few lines on this in the README.
In other words, nothing is wrong if there's an exception, it just means the user didn't write a new task_config. But I should add a warning.

hadasah replied:
We have a default task_config with the title, description, tags of the task to be shown on MTurk, but if the user writes one, I try to load it. I decided to go with having them write it in a .py file instead of .json (which would be easier to deal with) because I kept Jack's task_config.py template. There are a few lines on this in the README.
In other words, nothing is wrong if there's an exception, it just means the user didn't write a new task_config. But I should add a warning.

stephenroller commented:
We don't need to address all of those comments. we can chat irl about this

stephenroller commented on parlai/mturk/tasks/acute_eval/run.py:
```diff
@@ -0,0 +1,413 @@
+    specs['question'] = question
+    if pairing_dict['is_onboarding']:
+        specs['is_onboarding'] = True
+        onboarding_tasks[internal_pair_id] = task_data
```
Let’s file a new task to follow up on this

stephenroller commented on parlai/mturk/tasks/acute_eval/run.py:
```diff
@@ -0,0 +1,413 @@
+        tries += 1
+
+        internal_pair_id = next_task['task_specs']['internal_pair_id']
+        dialogue0_id = next_task['pairing_dict']['dialogue_dicts'][0]['id']
```
In the follow up task

stephenroller commented on parlai/mturk/tasks/acute_eval/run.py:
```diff
@@ -0,0 +1,413 @@
+    try:
+        import sys
+
+        sys.path.append(opt['task_folder'])
```
Explain this plz

hadasah replied:
lol I'll just copy what I said to emily:

"We have a default task_config with the title, description, tags of the task to be shown on MTurk, but if the user writes one, I try to load it. I decided to go with having them write it in a .py file instead of .json (which would be easier to deal with) because I kept Jack's task_config.py template. There are a few lines on this in the README.
In other words, nothing is wrong if there's an exception, it just means the user didn't write a new task_config. But I should add a warning."

I guess this is super nonstandard though based on your reaction. Maybe I should switch to a .json task_config file.

hadasah replied:
discussed offline, this is going into a command line arg

stephenroller commented:
Okay I explicitly marked follow ups with 👀 and never minds with 👎

hadasah commented on parlai/mturk/tasks/acute_eval/example_script.py:
```diff
@@ -0,0 +1,66 @@
+    args['max_hits_per_worker'] = 1  # max # hits a worker may complete
+    args['is_sandbox'] = True  # set to False to release real hits
+
+    args['annotations_per_pair'] = 1  # #times to use the same conversation pair
```
it was meant to be #times as in "number of times" but yeah that looks confusing at a glance

pringshia commented on parlai/mturk/tasks/acute_eval/frontend/components/custom.jsx:
```diff
@@ -0,0 +1,576 @@
+      this.handleResize();
+    }, 10);
+
+    top_pane_style["height"] = this.state.chat_height + "px";
```
any reason why this line isn't just part of the `top_pane_style` object above?

hadasah replied:
I don't remember there being a good reason, I can probalby move it next PR
